backend/FwLite/LcmCrdt/CrdtMiniLcmApi.cs

```diff
@@ -237,21 +282,97 @@ await dataModel.AddChanges(ClientId,
             ..await entry.Senses.ToAsyncEnumerable()
                 .SelectMany(s => CreateSenseChanges(entry.Id, s))
                 .ToArrayAsync(),
-            ..entry.Components.Select(c => new AddEntryComponentChange(c)),
-            ..entry.ComplexForms.Select(c => new AddEntryComponentChange(c)),
-            ..entry.ComplexFormTypes.Select(c => new AddComplexFormTypeChange(entry.Id, c))
+            ..await ToComplexFormComponents(entry.Components).ToArrayAsync(),
+            ..await ToComplexFormComponents(entry.ComplexForms).ToArrayAsync(),
+            ..await ToComplexFormTypes(entry.ComplexFormTypes).ToArrayAsync()
         ]);
         return await GetEntry(entry.Id) ?? throw new NullReferenceException();
+
+        async IAsyncEnumerable<AddEntryComponentChange> ToComplexFormComponents(IList<ComplexFormComponent> complexFormComponents)
+        {
+            foreach (var complexFormComponent in complexFormComponents)
+            {
+                if (complexFormComponent.ComponentEntryId == default) complexFormComponent.ComponentEntryId = entry.Id;
+                if (complexFormComponent.ComplexFormEntryId == default) complexFormComponent.ComplexFormEntryId = entry.Id;
+                if (complexFormComponent.ComponentEntryId == complexFormComponent.ComplexFormEntryId)
+                {
+                    throw new InvalidOperationException($"Complex form component {complexFormComponent} has the same component id as its complex form");
```

Comment:
All these exceptions are being thrown in `GetEntry` and are testing for things we don't expect to be possible right?
They're also slow, because they're behind db queries.

Shouldn't this validation essentially be dev-mode asserts?

Reply:
these are being thrown in `CreateEntry` I think the code folding in github is hiding the fact that `CreateEntry` is directly above not `GetEntry`. So in this case it's input validation which we obviously want at runtime. That said I am thinking about using a C# validation library for this as it's quite verbose

Reply:
My bad